src/guide/typescript-support.md

```diff
@@ -0,0 +1,270 @@
+        title: 'Vue 3 Guide',
+        author: 'Vue Team',
+        year: 2020
+      } as Book
+    }
+  }
+})
+```
+
+### Annotating Return Types
+
+Because of the circular nature of Vue’s declaration files, TypeScript may have difficulties inferring the types of computed. For this reason, you may need to annotate the return type computed properties.
+
+```ts
+import { defineComponent } from 'vue'
+
+const Component = defineComponent({
+  data() {
+    return {
+      message: 'Hello!'
+    }
+  },
+  computed: {
+    // needs an annotation
+    greeting(): string {
+      return this.message + '!'
+    }
+
+    // in a computed with a setter, getter needs to be annotated
+    greetingUppercased: {
+      get(): string {
+        return this.greeting.toUpperCase();
+      },
+      set(newValue: string) {
+        this.message = newValue.toUpperCase();
+      },
+    },
+  }
+})
+```
+
+### Annotating Props
+
+Vue does a runtime validation on props with a `type` defined. To provide these types to TypeScript, we need to cast the constructor with `PropType`:
+
+```ts
+import { defineComponent, PropType } from 'vue'
+
+interface ComplexMessage {
+  title: string
+  okMessage: string
+  cancelMessage: string
+}
+const Component = defineComponent({
+  props: {
+    name: String,
+    success: { type: String },
+    callback: {
+      type: Function as PropType<() => void>
+    },
+    message: {
+      type: Object as PropType<ComplexMessage>,
+      required: true,
+      validator(message: ComplexMessage) {
+        return !!message.title
+      }
+    }
+  }
+})
+```
+
+If you find validator not getting type inference or member completion isn’t working, annotating the argument with the expected type may help address these problems.
+
+## Using with Composition API
+
+On `setup()` function, you don't need to pass a typing to `props` parameter as it will infer types from `props` component option.
+
+```ts
+import { defineComponent } from 'vue'
+
+const Component = defineComponent({
+  props: {
+    message: {
+      type: String,
+      required: true
+    }
+  },
+
+  setup(props) {
+    const result = props.message.split('') // correct, 'message' is typed as a string
+    const filtered = props.message.filter(p => p.value) // an error will be thrown: Property 'filter' does not exist on type 'string'
+  }
+})
+```
+
+### Typing `ref`s
+
+To type a ref, we need to specify a type when calling `ref()` method:
+
+```ts
+import { defineComponent, ref } from 'vue'
+
+const Component = defineComponent({
+  setup() {
+    const year = ref<number>(2020)
+  }
+})
+```
+
+Sometimes we may need to specify complex types for a ref's inner value. We can do that succinctly by passing a generics argument when calling ref to override the default inference:
+
+```ts
+const year = ref<string | number>('2020') // year's type: Ref<string | number>
+
+year.value = 2020 // ok!
+```
+
+If the type of the generic is unknown, it's recommended to cast `ref` to `Ref<T>`:
```

Comment:
Not sure about this paragraph, the following example `State` is not `unknown`, I think it will make it better by just having, when using with generic is recommended to cast ...

Reply:
Removed an example 👍🏻